### PR TITLE
Introduce horizontal colorbars

### LINKED ISSUE #1244: allow for horizontal colorbar
_feature request:_ **horizontal colorbar** for heatmap, 3d graphs, etc.

The api for the colorbar seems to only allow for a colorbar positioned _vertically_.
image

But what if someone wants the bar displayed _horizontally_?
As far as I can tell from the api, this is not possible, but it would be nice to have.

image

### PR DESCRIPTION
**Pull request by archmoj**
Resolves #1244 by adding `orientation` attribute to `colorbar` with the option *'h'*.

@plotly/plotly_js

### CODE REVIEW

**Comment by nicolaskruchten** [comment]
Looks good to me! Automargins on the right and left side don't work the way I would expect, but other than that it looks great!

File: src/components/colorbar/attributes.js [hunk 1]
```diff
@@ -61,21 +57,23 @@ module.exports = overrideAll({
...
     x: {
         valType: 'number',
-        dflt: 1.02,
         min: -2,
         max: 3,
         description: [
-            'Sets the x position of the color bar (in plot fraction).'
+            'Sets the x position of the color bar (in plot fraction).',
+            'Defaults to 1.02 when `orientation` if *v* and ',
+            'defaults to 0.5 when `orientation` if *h*.'
```
Comment (alexcjohnson) [inline src/components/colorbar/attributes.js:65]:
```suggestion
            '0.5 when `orientation` if *h*.'
```

File: src/components/colorbar/attributes.js [hunk 2]
```diff
@@ -61,21 +57,23 @@ module.exports = overrideAll({
...
         valType: 'enumerated',
         values: ['left', 'center', 'right'],
-        dflt: 'left',
         description: [
             'Sets this color bar\'s horizontal position anchor.',
             'This anchor binds the `x` position to the *left*, *center*',
-            'or *right* of the color bar.'
+            'or *right* of the color bar.',
+            'Defaults to *left* when `orientation` if *v* and ',
+            'defaults to *center* when `orientation` if *h*.'
```
Comment (alexcjohnson) [inline src/components/colorbar/attributes.js:76]:
```suggestion
            'Defaults to *left* when `orientation` is *v* and',
            '*center* when `orientation` is *h*.'
```

File: src/components/colorbar/attributes.js (hunk 1 above)
Comment (alexcjohnson) [inline src/components/colorbar/attributes.js:65]:
```suggestion
            'Defaults to 1.02 when `orientation` is *v* and',
            '0.5 when `orientation` is *h*.'
```

File: src/components/colorbar/attributes.js [hunk 3]
```diff
@@ -86,21 +84,23 @@ module.exports = overrideAll({
...
     y: {
         valType: 'number',
-        dflt: 0.5,
         min: -2,
         max: 3,
         description: [
-            'Sets the y position of the color bar (in plot fraction).'
+            'Sets the y position of the color bar (in plot fraction).',
+            'Defaults to 1.02 when `orientation` if *v* and ',
+            'defaults to 0.5 when `orientation` if *h*.'
```
Comment (alexcjohnson) [inline src/components/colorbar/attributes.js:92]:
```suggestion
            'Defaults to 0.5 when `orientation` is *v* and',
            '1.02 when `orientation` is *h*.'
```

File: src/components/colorbar/attributes.js [hunk 4]
```diff
@@ -86,21 +84,23 @@ module.exports = overrideAll({
...
         valType: 'enumerated',
         values: ['top', 'middle', 'bottom'],
-        dflt: 'middle',
         description: [
             'Sets this color bar\'s vertical position anchor',
             'This anchor binds the `y` position to the *top*, *middle*',
-            'or *bottom* of the color bar.'
+            'or *bottom* of the color bar.',
+            'Defaults to *middle* when `orientation` if *v* and ',
+            'defaults to *bottom* when `orientation` if *h*.'
```
Comment (alexcjohnson) [inline src/components/colorbar/attributes.js:103]:
```suggestion
            'Defaults to *middle* when `orientation` is *v* and',
            '*bottom* when `orientation` is *h*.'
```

File: src/components/colorbar/attributes.js [hunk 5]
```diff
@@ -143,18 +143,22 @@ module.exports = overrideAll({
...
             'outside top', 'inside top',
+            'outside left', 'inside left',
+            'outside right', 'inside right',
             'outside bottom', 'inside bottom'
         ],
         dflt: 'outside',
-        description: [
-            'Determines where tick labels are drawn.'
-        ].join(' ')
+        description: 'Determines where tick labels are drawn.'
```
Comment (alexcjohnson) [inline src/components/colorbar/attributes.js:159]: Let's bring your comment above into the description, and clarify a little.
```suggestion
        description: [
            'Determines where tick labels are drawn relative to the ticks.',
            'Left and right options are used when `orientation` is *h*,',
            'top and bottom when `orientation` is *v*.'
        ].join(' ')
```

**Comment by alexcjohnson** [comment]
A couple of things I notice about how things are laid out:

Titles on the side: when the ticks & labels are on the inside it definitely seems like the title should be vertically centered. I think that also makes sense when the labels are outside (centered vertically wrt. the bar and labels together, not just the bar) but it's a little less clear to me in that case.
Screen Shot 2021-12-02 at 9 45 12 AM
Screen Shot 2021-12-02 at 9 47 32 AM

Outside labels on horizontal bars: first and last labels seem to intrude on the border padding. This is most evident with left/right `ticklabelposition` when the colorbar has a border and/or background color (as plenty of your excellent mocks do 🎉 ) but can also be seen with centered labels as in the second image above. I think without border & background the way it's drawn right now makes sense, as it would be unpleasant to have multiple colorbars that don't line up just because their tick labels have different lengths, but when you DO have a border or background color (different from `paper_bgcolor`) this situation should shrink the colorbar.
Screen Shot 2021-12-02 at 9 52 41 AM

**Comment by archmoj** [comment]
> when you DO have a border or background color (different from `paper_bgcolor`) this situation should shrink the colorbar. Screen Shot 2021-12-02 at 9 52 41 AM

Is it only for horizontal colorbars or both orientations?

**Comment by archmoj** [comment]
> Titles on the side: when the ticks & labels are on the inside it definitely seems like the title should be vertically centered. I think that also makes sense when the labels are outside (centered vertically wrt. the bar and labels together, not just the bar) but it's a little less clear to me in that case. Screen Shot 2021-12-02 at 9 45 12 AM Screen Shot 2021-12-02 at 9 47 32 AM
> 

This part was addressed in https://github.com/plotly/plotly.js/pull/6024/commits/8a137e3b9f1717312f49958b16a7b3087405e822.

**Comment by archmoj** [comment]
> Outside labels on horizontal bars: first and last labels seem to intrude on the border padding. 

This is actually not true. And I don't see a reason for a change to this PR. 
We should be good to move forward.

**Comment by archmoj** [comment]
> Outside labels on horizontal bars: first and last labels seem to intrude on the border padding. This is most evident with left/right `ticklabelposition` when the colorbar has a border and/or background color (as plenty of your excellent mocks do tada ) but can also be seen with centered labels as in the second image above. I think without border & background the way it's drawn right now makes sense, as it would be unpleasant to have multiple colorbars that don't line up just because their tick labels have different lengths, but when you DO have a border or background color (different from `paper_bgcolor`) this situation should shrink the colorbar. Screen Shot 2021-12-02 at 9 52 41 AM

This part is addressed in https://github.com/plotly/plotly.js/pull/6024/commits/2f09b6e06ee0baf33338b449b833be3bac6dd126 by adjusting the position of first and last tick labels.

**Comment by alexcjohnson** [comment]
How about we just remove these extremal ticks, in the cases you determine you need to shift them? It's an edge case already since mostly people won't be putting their colorbars in boxes, and the shifted labels are too easy to break (and even some non-broken cases don't look that great, when the number intersects the tick)
Screen Shot 2021-12-10 at 12 01 50 AM
Screen Shot 2021-12-10 at 12 02 37 AM
Screen Shot 2021-12-10 at 12 03 03 AM

Also, and this one I'm comfortable just making an issue for and deferring, `outside left` with many labels such that they auto-rotate does weird things:
Screen Shot 2021-12-10 at 12 10 55 AM
Screen Shot 2021-12-10 at 12 10 28 AM

This is an issue for regular cartesian axes too
Screen Shot 2021-12-10 at 12 15 54 AM

and for all shifted tick labels on both colorbars and cartesian axes if you provide explicit `tickangle`

File: src/components/colorbar/draw.js [hunk 6]
```diff
@@ -697,6 +698,55 @@ function drawColorBar(g, opts, gd) {
             gs.t - (isVertical ? 0 : (1 - yRatio) * outerThickness - moveY)
         ));
 
+        if(!isVertical && (
+            borderwidth || (
+                tinycolor(bgcolor).getAlpha() &&
+                !tinycolor.equals(fullLayout.plot_bgcolor, bgcolor)
```
Comment (alexcjohnson) [inline src/components/colorbar/draw.js:704]: `paper_bgcolor`?

**Comment by archmoj** [comment]
> How about we just remove these extremal ticks, in the cases you determine you need to shift them? 

No I think it's better to keep them around and adjust their position. 
Otherwise no labels would show up here:

h-colorbar_geo_multiple-usa-choropleths

**Comment by alexcjohnson** [comment]
OK, but it's way too easy to break and the breaking cases I noted above are just the tip of the iceberg - this would be a huge pain to fix. How about this then: as long as there will be at least 2 labels left, remove the ones that would need moving; if removing would result in 0 or 1 labels, we move them instead.

**Comment by alexcjohnson** [comment]
> > Titles on the side: when the ticks & labels are on the inside it definitely seems like the title should be vertically centered. I think that also makes sense when the labels are outside (centered vertically wrt. the bar and labels together, not just the bar) but it's a little less clear to me in that case. Screen Shot 2021-12-02 at 9 45 12 AM Screen Shot 2021-12-02 at 9 47 32 AM
> 
> This part was addressed in [8a137e3](https://github.com/plotly/plotly.js/commit/8a137e3b9f1717312f49958b16a7b3087405e822).

8a137e3 adjusted the horizontal alignment - I was talking about vertical

**Comment by archmoj** [comment]
> OK, but it's way too easy to break and the breaking cases I noted above are just the tip of the iceberg - this would be a huge pain to fix. How about this then: as long as there will be at least 2 labels left, remove the ones that would need moving; if removing would result in 0 or 1 labels, we move them instead.

Applied in https://github.com/plotly/plotly.js/pull/6024/commits/07cd536772decbc1ba12d88a18c9a07a72bcf7aa

**Comment by archmoj** [comment]
> > > Titles on the side: when the ticks & labels are on the inside it definitely seems like the title should be vertically centered. I think that also makes sense when the labels are outside (centered vertically wrt. the bar and labels together, not just the bar) but it's a little less clear to me in that case. Screen Shot 2021-12-02 at 9 45 12 AM Screen Shot 2021-12-02 at 9 47 32 AM
> > 
> > 
> > This part was addressed in [8a137e3](https://github.com/plotly/plotly.js/commit/8a137e3b9f1717312f49958b16a7b3087405e822).
> 
> [8a137e3](https://github.com/plotly/plotly.js/commit/8a137e3b9f1717312f49958b16a7b3087405e822) adjusted the horizontal alignment - I was talking about vertical

The vertical alignment based on the height of colorbar title could be an option later on the road. 
If one looks at the whole graph not only the colorbar, the current setup is pretty good IMHO.

**Review by alexcjohnson — APPROVED** [review]
Excellent - let's add one or two mocks that exactly match existing mocks but with no border or bg for the colorbar, showing that the labels are all visible and unmoved. Then I'd say this is ready to merge and release! 💃

**Comment by archmoj** [comment]
> Excellent - let's add one or two mocks that exactly match existing mocks but with no border or bg for the colorbar, showing that the labels are all visible and unmoved. Then I'd say this is ready to merge and release! woman_dancing

Good call. Done in https://github.com/plotly/plotly.js/pull/6024/commits/a0901bd063c004c27e9fd9e7df83fbb578e978de.

**Comment by etpinard** [comment]
Great feature! 

Well done @archmoj !!

**Comment by archmoj** [comment]
> Great feature!
> 
> Well done @archmoj !!

Thanks for the review @etpinard :)

**Pull request merged**